File: JavaExamples/src/main/java/org/firstinspires/ftc/teamcode/examplerobot/Shooter.java
package org.firstinspires.ftc.teamcode.examplerobot;

import androidx.annotation.NonNull;

import com.pedropathing.localization.MotionState;
import com.pedropathing.math.Pose;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.jetbrains.annotations.NotNull;

import dev.frozenmilk.dairy.mercurial.processes.Channel;
import dev.frozenmilk.dairy.mercurial.processes.EventManager;
import dev.frozenmilk.dairy.mercurial.processes.EventManager.EventHandled;
import dev.frozenmilk.dairy.mercurial.processes.StateMachine;

public class Shooter extends StateMachine {
    private final DcMotorEx motor;

    public Shooter(
        HardwareMap hardwareMap,
        EventManager<MotionState> localizationEventManager
    ) {
        motor = hardwareMap.get(DcMotorEx.class, "shooter");
        motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        // we tell the localiser to send events to us
        localizationEventManager.addHandler(motionStateForwarder);
    }

    // the channel we will receive motion states into
    private final Channel<MotionState> motionStates = Channel.single();

    // we need the ability to switch modes
    private final Channel<Mode> switches = Channel.single();

    // a fixed power
    private class Fixed implements Mode {
        public final double power;

        public Fixed(double power) {
            this.power = power;
        }

        @Override
        public void enter(@NotNull Mode previousMode) {
            motor.setPower(power);
        }

        @Override
        @NotNull
        public Mode eval() {
            Mode nextMode = switches.poll();
            if (nextMode == null) return this;
            else return nextMode;
        }
    }

    // idle
    private final Mode idle = new Mode() {
        @Override
        public void enter(@NotNull Mode previousMode) {
            motor.setPower(0);
        }

        @Override
        @NotNull
        public Mode eval() {
            Mode nextMode = switches.poll();
            if (nextMode == null) return this;
            else return nextMode;
        }
    };

    // calculate the power from the position of the robot and the target position
    private class Target implements Mode {
        public final Pose target;

        private Target(Pose target) {
            this.target = target;
        }

        @Override
        @NotNull
        public Mode eval() {
            MotionState motionState = motionStates.poll();
            if (motionState != null) {
                motor.setPower(
                    0.1 + 0.001 * Math.hypot(
                        target.x() - motionState.pose().x(),
                        target.y() - motionState.pose().y()
                    )
                );
            }
            Mode nextMode = switches.poll();
            if (nextMode == null) return this;
            else return nextMode;
        }
    }

    // this event handler will forward motion state events to the shooter
    private final EventManager.Handler<MotionState> motionStateForwarder = motionState -> {
        motionStates.send(motionState);
        // if the shooter finishes, remove this handler
        if (fiber().status().alive()) return EventHandled.ok();
        else return EventHandled.remove();
    };

    @Override
    @NotNull
    protected Mode init() {
        // we init into the idle mode
        return idle;
    }

    // public api:
    public void target(Pose pose) {
        switches.send(new Target(pose));
    }

    public void idle() {
        switches.send(idle);
    }

    public void fixed(double power) {
        switches.send(new Fixed(power));
    }
}

File: KotlinExamples/src/main/kotlin/org/firstinspires/ftc/teamcode/examplerobot/Shooter.kt
package org.firstinspires.ftc.teamcode.examplerobot

import com.pedropathing.localization.MotionState
import com.pedropathing.math.Pose
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.HardwareMap
import dev.frozenmilk.dairy.mercurial.processes.Channel
import dev.frozenmilk.dairy.mercurial.processes.Channel.Companion.tryPoll
import dev.frozenmilk.dairy.mercurial.processes.EventManager
import dev.frozenmilk.dairy.mercurial.processes.StateMachine
import kotlin.math.hypot

class Shooter(
    hardwareMap: HardwareMap,
    localizationEventManager: EventManager<MotionState>,
) : StateMachine() {
    private val motor = hardwareMap[DcMotorEx::class.java, "shooter"]

    init {
        motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
    }

    // the channel we will receive motion states into
    private val motionStates = Channel.single<MotionState>()

    // we need the ability to switch modes
    private val switches = Channel.single<Mode>()

    // a fixed power
    private inner class Fixed(val power: Double) : Mode {
        override fun enter(previousMode: Mode) {
            motor.power = power
        }

        override fun eval(): Mode = switches.poll() ?: this
    }

    // idle
    private val idle = object : Mode {
        override fun enter(previousMode: Mode) {
            motor.power = 0.0
        }

        override fun eval(): Mode = switches.poll() ?: this
    }

    // calculate the power from the position of the robot and the target position
    private inner class Target(val target: Pose) : Mode {
        override fun eval(): Mode {
            motionStates.tryPoll { motionState ->
                motor.power = 0.1 + 0.001 * hypot(
                    target.x() - motionState.pose().x(),
                    target.y() - motionState.pose().y(),
                )
            }
            return switches.poll() ?: this
        }
    }

    // this event handler will forward motion state events to the shooter
    private val motionStateForwarder = object : EventManager.Handler<MotionState> {
        override fun handleEvent(event: MotionState): EventManager.EventHandled<MotionState> {
            motionStates.send(event)
            // if the shooter finishes, remove this handler
            return if (fiber.status.alive) EventManager.EventHandled.Ok
            else EventManager.EventHandled.Remove
        }
    }

    init {
        // we tell the localiser to send events to us
        localizationEventManager.addHandler(motionStateForwarder)
    }

    override fun init(): Mode {
        // we init into the idle mode
        return idle
    }

    // public api:
    fun target(pose: Pose) {
        switches.send(Target(pose))
    }

    fun idle() {
        switches.send(idle)
    }

    fun fixed(power: Double) {
        switches.send(Fixed(power))
    }
}